gr-osmosdr 0.1.4 (new formula)

mikemorris opened the pull request:
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

Add new formula for Osmocom's (Open Source Mobile Communication) blocks for GNU Radio which enable to use several SDRs (Software Defined Radio) like the famous low-cost RTL-SDR, HackRF, USRP and the like.

Refs https://github.com/Homebrew/legacy-homebrew/pull/50489

Both `0.1.4` and `--HEAD` (currently my fork with a tiny patch applied, which I need to submit upstream) build successfully on my macOS 10.13.2 machine. There are probably some implicit dependencies that I already have installed which will need to be added. This doesn't have a test yet, and I'm not quite sure what would be useful to add for a library with no binaries.

/cc @muellermartin

muellermartin commented on Formula/gr-osmosdr.rb:
```diff
@@ -0,0 +1,19 @@
+  url "https://cgit.osmocom.org/gr-osmosdr/snapshot/gr-osmosdr-0.1.4.tar.xz"
+  sha256 "1945d0d98fd4b600cb082970267ec2041528f13150422419cbd7febe2b622721"
+  # head "git://git.osmocom.org/gr-osmosdr"
+  head "git://github.com/mikemorris/gr-osmosdr.git"
```
Why did you provide your fork of gr-osmosdr as repository for the head version instead of the official one (or its GitHub mirror)?

fxcoudert replied:
Remove head

mikemorris replied:
I added my fork because of a small patch required for it to compile with clang-9 (default on macOS 10.13.2, haven't tested with other compilers) - https://github.com/mikemorris/gr-osmosdr/commit/ae9bd438c31d5c519cd0874027952f11bdba9cf4

Is there a way to move this patch into the Homebrew formula to keep the official repository as `head`?

mikemorris replied:
Figured out how to add the patch in formula, hopefully it will be accepted upstream though.

muellermartin commented:
Great that you try to bring gr-osmosdr to Homebrew!

You mentioned that there might be some implicit dependencies and I guess you mean the individual libraries for each supported hardware (e.g. RTL-SDR, HackRF, AirSpy). I highly recommend to add at least `librtlsdr` as recommended dependency as most people will likely have a RTL-SDR in use. Further hardware might then be added step by step in the future.

fxcoudert commented on Formula/gr-osmosdr.rb:
```diff
@@ -0,0 +1,19 @@
+    system "make", "install"
+  end
+
+  test do
```
A good test would be a simple source program, compiling, linking and running against the newly built library.

fxcoudert commented on Formula/gr-osmosdr.rb:
```diff
@@ -0,0 +1,19 @@
+class GrOsmosdr < Formula
+  desc "Osmocom Gnu Radio Blocks"
```
`GNU` uppercase

mikemorris commented:
Addressed a few of the comments, added optional hardware libraries and Python support with swig. Working on the test, but I'm hitting an error when building with `--HEAD --with-librtlsdr`

```
cd /tmp/gr-osmosdr-20171229-47611-1towtii/lib && /usr/local/Homebrew/Library/Homebrew/shims/super/clang++  -DBOOST_ALL_DYN_LINK -DHAVE_CONFIG_H=1 -DHAVE_FCD=1 -DUSE_SSE2 -Dgnuradio_osmosdr_EXPORTS -I/tmp/gr-osmosdr-20171229-47611-1towtii/lib -I/tmp/gr-osmosdr-20171229-47611-1towtii/include -I/usr/local/include -I/tmp/gr-osmosdr-20171229-47611-1towtii/lib/fcd -I/tmp/gr-osmosdr-20171229-47611-1towtii/lib/file -I/tmp/gr-osmosdr-20171229-47611-1towtii/lib/rtl -I/tmp/gr-osmosdr-20171229-47611-1towtii/lib/rtl_tcp -I/tmp/gr-osmosdr-20171229-47611-1towtii/lib/uhd -I/tmp/gr-osmosdr-20171229-47611-1towtii/lib/rfspace -I/tmp/gr-osmosdr-20171229-47611-1towtii/lib/redpitaya  -DNDEBUG -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.13.sdk -fPIC   -Wall -Wextra -Wno-unused-parameter -Wsign-compare -fvisibility=hidden -fvisibility-inlines-hidden -msse2 -o CMakeFiles/gnuradio-osmosdr.dir/rtl_tcp/rtl_tcp_source_c.cc.o -c /tmp/gr-osmosdr-20171229-47611-1towtii/lib/rtl_tcp/rtl_tcp_source_c.cc
/tmp/gr-osmosdr-20171229-47611-1towtii/lib/rtl/rtl_source_c.cc:224:9: error: use of undeclared identifier 'rtlsdr_set_bias_tee'
  ret = rtlsdr_set_bias_tee(_dev, bias_tee);
        ^
1 error generated.
make[2]: *** [lib/CMakeFiles/gnuradio-osmosdr.dir/rtl/rtl_source_c.cc.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [lib/CMakeFiles/gnuradio-osmosdr.dir/all] Error 2
make: *** [all] Error 2
```

muellermartin commented:
I think the error regarding the missing identifier related to librtlsdr is due to the missing dependency of librtlsdr. Try adding `depends_on "librtlsdr"` to the Formula.

mikemorris commented:
Figured it out, `HEAD` of `gr-osmosdr` just needs to be built against `HEAD` of `librtlsdr`. Think all that's left is finishing up this test and checking that all of the build options compile successfully.

mikemorris commented on Formula/gr-osmosdr.rb:
```diff
@@ -0,0 +1,101 @@
+    url "git://git.osmocom.org/gr-osmosdr"
+
+    if build.with?("librtlsdr")
+      opoo "Building --HEAD --with-librtlsdr requires `brew install librtlsdr --HEAD`"
```
@tdsmith I'm hitting precisely the issue in https://github.com/Homebrew/legacy-homebrew/issues/17387#issuecomment-91925218 with the `librtlsdr` dependency.

mikemorris commented on Formula/gr-osmosdr.rb:
```diff
@@ -0,0 +1,100 @@
+  depends_on "python" => :optional
+
+  if build.head? && build.with?("librtlsdr")
+    opoo "Building --HEAD --with-librtlsdr requires `brew install librtlsdr --HEAD`"
```
@tdsmith I'm hitting precisely the issue in https://github.com/Homebrew/legacy-homebrew/issues/17387#issuecomment-91925218 with the `librtlsdr` dependency.

mikemorris commented:
Think this is close to being ready, successfully builds `--with-python` and disables Python support when the flag isn't set and the only warning left in `brew audit --new-formula gr-osmosdr` is regarding the patch that I've submitted upstream but haven't heard anything back yet.

ilovezfs commented:
@BrewTestBot test this please

ilovezfs commented:
@BrewTestBot test this please

mikemorris commented:
@ilovezfs So, each of the options (aside from `--with-python` for Python bindings) are for adding support for a single hardware SDR dongle - many users would only need to enable one for the specific device they're using, so it doesn't particularly make sense to require them all by default.

mikemorris commented:
@ilovezfs I can remove `head` (and the corresponding patch and `librtlsdr` warning), just curious as to why it should be excluded? `gr-osmosdr` has had bugfixes and ongoing development happening in the master branch, but the last actual release tag was ~3 years ago.

ilovezfs commented:
>but the last actual release tag was ~3 years ago.

yikes. that sounds like a pretty good reason not to have a formula

stromnet commented:
Thanks for Formula!

I got this during build:
```
$ brew install gr-osmosdr --with-python --with-librtlsdr --with-airspy
...
==> python -c import setuptools... --no-user-cfg install --prefix=/usr/local/Cellar/gr-osmosdr/0.1.4/libexec/vendor --single-version-externally-managed --record=installed.txt
==> cmake . -DCMAKE_C_FLAGS_RELEASE=-DNDEBUG -DCMAKE_CXX_FLAGS_RELEASE=-DNDEBUG -DCMAKE_INSTALL_PREFIX=/usr/local/Cellar/gr-osmosdr/0.1.4 -DCMAKE_BUILD_TYPE=Release -DCMAKE_FIND_FRAMEWORK=LAST -DCMAKE_VERBOSE_MAKEFILE=ON -Wno-dev -DHAVE_CLOCK
==> make install
Warning: gr-osmosdr dependency gcc was built with a different C++ standard
library (libstdc++ from clang). This may cause problems at runtime.
🍺  /usr/local/Cellar/gr-osmosdr/0.1.4: 154 files, 3.2MB, built in 37 seconds
```

It seems to work though:
```
$ python -c 'import gnuradio, osmosdr';
Mac OS; Clang version 8.0.0 (clang-800.0.42.1); Boost_106600; UHD_003.010.002.000-0-unknown
```

And ham2mon works fine

muellermartin commented:
Bump

muellermartin commented:
Bump²

What's the state of this formula? Can the options remain (as they are important for the hardware support)? And should the head be removed?

Is there anything I can help (e.g. testing or providing a patch)?

fxcoudert commented:
@muellermartin For this to be suitable for Homebrew core inclusion, we will need you to:
- remove head
- determine a "most commonly used" set of dependencies, and make them mandatory, getting rid of all options; we strive to provide feature-complete binary packages

We know that our core policies are not ideal for all software and distribution cases, which is why we make it easy to maintain the formula in a personal tap: https://github.com/Homebrew/brew/blob/master/docs/How-to-Create-and-Maintain-a-Tap.md This is probably the best option if the community wants to have a compile-from-source formula, integrated within the Homebrew framework, with lots of options and possibly using non-released versions.

muellermartin commented:
Thanks for the clarification! I'm trying to change the formula to fit Homebrew's needs and submit it to @mikemorris branch. Currently there seems to be a minor issue with libbladerf (CMake fails with an error `A duplicate ELSE command was found inside an IF block` which I'm investigating)...

fxcoudert commented:
@muellermartin libbladerf has been a pain: it has never built in Homebrew since Sierra, and I'd be tempted to remove it altogether unless upstream decides to release a new version that fixes the problem

muellermartin commented:
Good to know. This issue really seems to be minor (it seems to be related to installing udev rules which are not needed on macOS anyways so maybe I can submit an upstream patch).

fxcoudert commented:
@muellermartin the cmake issue fixed by https://github.com/Nuand/bladeRF/commit/037e2886
but there is another one with `clockid_t` :(

fxcoudert commented:
@muellermartin libbladerf is being fixed https://github.com/Homebrew/homebrew-core/pull/25111

muellermartin commented:
For some reason GitHub does not allow me to create a PR into the fork of @mikemorris so here is a manual PR:

```diff
From 6ca80a689d7385adf73963e08a3c78a47e8889f1 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Martin=20M=C3=BCller?= <mail@mueller-martin.net>
Date: Sun, 11 Mar 2018 16:43:36 +0100
Subject: [PATCH] gr-osmosdr: Remove head and make options mandatory

---
 Formula/gr-osmosdr.rb | 84 +++++++++------------------------------------------
 1 file changed, 14 insertions(+), 70 deletions(-)

diff --git a/Formula/gr-osmosdr.rb b/Formula/gr-osmosdr.rb
index 5825177dec..21df7616f5 100644
--- a/Formula/gr-osmosdr.rb
+++ b/Formula/gr-osmosdr.rb
@@ -4,49 +4,29 @@ class GrOsmosdr < Formula
   url "https://cgit.osmocom.org/gr-osmosdr/snapshot/gr-osmosdr-0.1.4.tar.xz"
   sha256 "1945d0d98fd4b600cb082970267ec2041528f13150422419cbd7febe2b622721"
 
-  head do
-    url "git://git.osmocom.org/gr-osmosdr"
-    patch :DATA
-  end
-
   depends_on "cmake" => :build
+  depends_on "swig" => :build
   depends_on "boost"
   depends_on "gnuradio"
-  depends_on "librtlsdr" => :optional
-  depends_on "hackrf" => :optional
-  depends_on "libbladerf" => :optional
-  depends_on "uhd" => :optional
-  depends_on "airspy" => :optional
-  depends_on "python" => :optional
-
-  if build.head? && build.with?("librtlsdr")
-    opoo "Building --HEAD --with-librtlsdr requires `brew install librtlsdr --HEAD`"
-  end
-
-  if build.with?("python")
-    depends_on "swig" => :build
-
-    resource "Cheetah" do
-      url "https://files.pythonhosted.org/packages/source/C/Cheetah/Cheetah-2.4.4.tar.gz"
-      sha256 "be308229f0c1e5e5af4f27d7ee06d90bb19e6af3059794e5fd536a6f29a9b550"
-    end
+  depends_on "librtlsdr"
+  depends_on "hackrf"
+  depends_on "uhd"
+  depends_on "airspy"
+  depends_on "python"
+
+  resource "Cheetah" do
+    url "https://files.pythonhosted.org/packages/source/C/Cheetah/Cheetah-2.4.4.tar.gz"
+    sha256 "be308229f0c1e5e5af4f27d7ee06d90bb19e6af3059794e5fd536a6f29a9b550"
   end
 
   def install
-    cmake_args = std_cmake_args
+    ENV.prepend_create_path "PYTHONPATH", libexec/"vendor/lib/python2.7/site-packages"
 
-    if build.with?("python")
-      ENV.prepend_create_path "PYTHONPATH", libexec/"vendor/lib/python2.7/site-packages"
-
-      resource("Cheetah").stage do
-        system "python", *Language::Python.setup_install_args(libexec/"vendor")
-      end
-    else
-      ohai "Building without Python support"
-      cmake_args << "-DENABLE_PYTHON=OFF"
+    resource("Cheetah").stage do
+      system "python", *Language::Python.setup_install_args(libexec/"vendor")
     end
 
-    system "cmake", ".", *cmake_args
+    system "cmake", ".", *std_cmake_args
     system "make", "install"
   end
 
@@ -62,39 +42,3 @@ class GrOsmosdr < Formula
     system "./test"
   end
 end
-
-__END__
-diff --git a/lib/rtl_tcp/rtl_tcp_source_c.cc b/lib/rtl_tcp/rtl_tcp_source_c.cc
-index ecdeee0..c962dd8 100644
---- a/lib/rtl_tcp/rtl_tcp_source_c.cc
-+++ b/lib/rtl_tcp/rtl_tcp_source_c.cc
-@@ -298,17 +298,17 @@ rtl_tcp_source_c::rtl_tcp_source_c(const std::string &args) :
-   // set direct sampling
-   struct command cmd;
- 
--  cmd = { 0x09, htonl(direct_samp) };
-+  cmd = (struct command){ 0x09, htonl(direct_samp) };
-   send(d_socket, (const char*)&cmd, sizeof(cmd), 0);
-   if (direct_samp)
-     _no_tuner = true;
- 
-   // set offset tuning
--  cmd = { 0x0a, htonl(offset_tune) };
-+  cmd = (struct command){ 0x0a, htonl(offset_tune) };
-   send(d_socket, (const char*)&cmd, sizeof(cmd), 0);
- 
-   // set bias tee
--  cmd = { 0x0e, htonl(bias_tee) };
-+  cmd = (struct command){ 0x0e, htonl(bias_tee) };
-   send(d_socket, (const char*)&cmd, sizeof(cmd), 0);
- }
- 
-@@ -567,7 +567,7 @@ bool rtl_tcp_source_c::set_gain_mode( bool automatic, size_t chan )
-   send(d_socket, (const char*)&cmd, sizeof(cmd), 0);
- 
-   // AGC mode
--  cmd = { 0x08, htonl(automatic) };
-+  cmd = (struct command){ 0x08, htonl(automatic) };
-   send(d_socket, (const char*)&cmd, sizeof(cmd), 0);
- 
-   _auto_gain = automatic;
-- 
2.16.2

```

mikemorris commented:
Thanks for your help on this @fxcoudert and @muellermartin, I can get that patch applied to my this branch tonight!

muellermartin commented:
Damn, the build failed because there seems to be an issue with Osmocom's installation of cgit, which currently is unable to generate the snapshot tarball. I'd suggest adding the official mirror on GitHub to the formula to fix this:

```patch
From c8467443242e9ceb553cfc631c136586b865e72c Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Martin=20M=C3=BCller?= <mail@mueller-martin.net>
Date: Mon, 2 Apr 2018 20:58:58 +0200
Subject: [PATCH] gr-osmosdr: Add mirror to fix unreachable snapshot

---
 Formula/gr-osmosdr.rb | 1 +
 1 file changed, 1 insertion(+)

diff --git a/Formula/gr-osmosdr.rb b/Formula/gr-osmosdr.rb
index 21df7616f5..be455d867c 100644
--- a/Formula/gr-osmosdr.rb
+++ b/Formula/gr-osmosdr.rb
@@ -2,6 +2,7 @@ class GrOsmosdr < Formula
   desc "Osmocom GNU Radio Blocks"
   homepage "https://osmocom.org/projects/sdr/wiki/GrOsmoSDR"
   url "https://cgit.osmocom.org/gr-osmosdr/snapshot/gr-osmosdr-0.1.4.tar.xz"
+  mirror "https://github.com/osmocom/gr-osmosdr/archive/v0.1.4.tar.gz"
   sha256 "1945d0d98fd4b600cb082970267ec2041528f13150422419cbd7febe2b622721"
 
   depends_on "cmake" => :build
-- 
2.16.3

```

muellermartin commented:
The last build failed at the audit step because the dependencies were not ordered alphabetically.

Following PR fixes this:

```patch
From c99cf0b946506ba5a7e2e10ce1e83ac18abf0b53 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Martin=20M=C3=BCller?= <mail@mueller-martin.net>
Date: Tue, 3 Apr 2018 01:07:40 +0200
Subject: [PATCH] gr-osmosdr: Change dependency order

---
 Formula/gr-osmosdr.rb | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/Formula/gr-osmosdr.rb b/Formula/gr-osmosdr.rb
index db8e8c3c12..b34ec56b57 100644
--- a/Formula/gr-osmosdr.rb
+++ b/Formula/gr-osmosdr.rb
@@ -7,13 +7,13 @@ class GrOsmosdr < Formula
 
   depends_on "cmake" => :build
   depends_on "swig" => :build
+  depends_on "airspy"
   depends_on "boost"
   depends_on "gnuradio"
-  depends_on "librtlsdr"
   depends_on "hackrf"
-  depends_on "uhd"
-  depends_on "airspy"
+  depends_on "librtlsdr"
   depends_on "python"
+  depends_on "uhd"
 
   resource "Cheetah" do
     url "https://files.pythonhosted.org/packages/source/C/Cheetah/Cheetah-2.4.4.tar.gz"
-- 
2.16.3

```

muellermartin commented:
Cool, the last build succeeded \o/ Thank you very much @mikemorris for applying my patches so quickly. Nice teamwork ;)

Now it's up to the Homebrew maintainers to include this formula. @ilovezfs or @fxcoudert could you please take a look?
